GH Actions: improve workflows

jrfnl opened the pull request:
## Commit Details

### GH Actions workflows: fix PHP set up step

To disable code coverage, the `coverage` key should be set to `none`. `false` is not a valid (or supported) value, so the behaviour may be unpredictable.

Ref: https://github.com/shivammathur/setup-php#disable-coverage

### GH Actions: run the tests against all supported PHP versions

The original Travis script ran the following commands against *all* supported PHP versions, but didn't test the phar:
```yaml
  - composer test
  - ./parallel-lint --exclude vendor --exclude tests/examples --no-colors .
  - ./parallel-lint --exclude vendor --exclude tests/examples .
```

The current `test` job only ran the unit tests against PHP 5.4. It did not run the integration test (running Parallel Lint against its own code)  via a direct call to the script anymore at all.
The integration test was basically now _only_ being run for the phar and only in one flavour, though it did do that on all supported PHP versions.

This commit:
* Removes the job which only ran the unit tests against PHP 5.4.
* Adds the running of the above three commands (unit tests and two versions of integrations tests runs via a direct call to the script), to the job which also runs the integration tests via the phar file.
* Updates the command line parameters used for the phar run to be the same as those used for the direct script call integration tests.

### GH Actions: actually get the tests running on all PHP versions

This fixes the problem originally identified by @roelofr with the `Fatal error: Interface 'JsonSerializable' not found` error.

The problem had nothing to do with the PHP setup, but everything with some inane setting of the Nette testing framework as can be seen in the transcript of older, failing trial runs done by @roelofr while setting up the GH Actions workflows:
```
Run composer test
> @php vendor/bin/tester -p php tests
 _____ ___  ___ _____ ___  ___
|_   _/ __)( __/_   _/ __)| _ )
  |_| \___ /___) |_| \___ |_|_\  v2.3.5

Note: No php.ini is used.
```

The key here is the `Note: No php.ini is used.`.

As the system `php.ini` was not being used, no extensions were loaded, which was causing the errors.

This is a change which was introduced in Nette Tester 2.0.0. As of that version, you need to always tell the Nette testing framework explicitly which `php.ini` file to use, `-C` tells it to use the system-wide `php.ini`, with `-c` you can specify a path to a `php.ini` file and by default **_no `php.ini` is used_**.
(_honestly, why did anyone ever think making that the default was a good idea ?!!?_)

As the tests will be running on different versions of the Nette Framework, Nette 1.x for PHP 5.4 and 5.5 and Nette 2.x for PHP 5.6+ and Nette 1.x, does not yet support the `-C` option and breaks when it is used, I've added a second test script to the `composer.json` file with the correct command line options to run Nette on PHP 5.4/5.5 and added conditions to the GH Actions workflow to use the correct script depending on the PHP/Nette test framework version being used.

Refs:
* https://tester.nette.org/en/running-tests#toc-c-path
* https://tester.nette.org/en/guide#toc-supported-php-versions
* https://github.com/nette/tester/releases/tag/v2.0.0

### GH Actions: allow for manually triggering a workflow

Triggering a workflow for a branch manually is not supported by default in GH Actions, but has to be explicitly allowed.

This is useful if, for instance, an external action script or composer dependency has broken.
Once a fix is available, failing builds for open PRs can be retriggered manually instead of having to be re-pushed to retrigger the workflow.

Ref: https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/

### GH Actions: report CS violations in the PR

The cs2pr tool will allow to display the results from an action run in checkstyle format in-line in the PR code view, which should improve usability of the workflow results.

This implements this for the style check command.

Includes switching the PHP version to PHP 7.4, as PHP 8.0 is not fully supported yet in PHP_CodeSniffer (full support expected in PHPCS 3.6.0).

Ref: https://github.com/staabm/annotate-pull-request-from-checkstyle

### GH Actions: report linting violations in the PR

The cs2pr tool will allow to display the results from an action run in checkstyle format in-line in the PR code view, which should improve usability of the workflow results.

This implements this for the results of running Parallel Lint over its own code.

This also implicitly acts as an integration test for the `checkstyle` output and the cs2pr integration.

Ref: https://github.com/staabm/annotate-pull-request-from-checkstyle

### GH Actions: clean up

* Remove the now redundant Travis script.
* Remove the reference to the Travis script from `.gitattributes`.
* Update the "Build status" badge in the README to show the result of the GH Actions run instead of Travis.

roelofr commented on .github/workflows/test.yml:
```diff
@@ -113,15 +99,33 @@ jobs:
+        if: ${{ matrix.php >= 5.6 }}
+        run: composer test
+
+      - name: 'Integration test 1 - linting own code, no colors'
+        run: ./parallel-lint --exclude vendor --exclude tests/examples --no-colors .
+
+      - name: 'Integration test 2 - linting own code'
+        run: ./parallel-lint --exclude vendor --exclude tests/examples . --checkstyle | cs2pr
```
The violations will not be reported if the first integration check fails, but I guess you'd want this to show in the report too right?

Maybe, we can always run integration test 2 if integration test 1 fails or passes, but not if it's skipped (failure early on). I'm guessing this might work.

```suggestion
      - name: 'Integration test 1 - linting own code, no colors'
        id: integration-test-1
        run: ./parallel-lint --exclude vendor --exclude tests/examples --no-colors .

      - name: 'Integration test 2 - linting own code'
        if: ${{ steps.integration-test-1.conclusion == 'failure' || steps.integration-test-1.conclusion == 'success' }}
        run: ./parallel-lint --exclude vendor --exclude tests/examples . --checkstyle | cs2pr
```

roelofr reviewed (commented):
Wow, thanks for finding that Nette thing, I spent far too much time missing that one lousy line.

I had one suggestion to always report linting status to PRs, but also print it in the actions pipeline (so we can see if the `--colors` works too).

glensc commented on .github/workflows/test.yml:
```diff
@@ -113,15 +99,33 @@ jobs:
+        run: ./parallel-lint --exclude vendor --exclude tests/examples --no-colors .
+
+      - name: 'Integration test 2 - linting own code'
+        run: ./parallel-lint --exclude vendor --exclude tests/examples . --checkstyle | cs2pr
```
if the lint own code fails, you won't get cs2pr results to pull the request as the erroring job cancels other jobs. you probably want to move cs2pr filtered job up.

And due to `| ` filtering, the second command won't even trigger the error code to shell.

```sh
➔ sh -c 'false; echo $?'
1
➔ sh -c 'false | cat; echo $?'
0
```

note, cs2pr itself may trigger error if it finds error in input, haven't checked that.

so if you fix the workflow based on that, be sure include the reasonning in the .yaml as a comment, very unlikely somebody will find this comment box when refactoring that part in there.
maybe just link to this post.

roelofr replied:
`cs2pr` has an option that insinuates the script exits if it finds any errors or warnings, so we should be good with the pipe.

> `--graceful-warnings`: Don't exit with error codes if there are only warnings

glensc replied:
yeah, but please read the first part of my message: but no reports to github. need still to switch execution order to get reports to github ui.

roelofr replied:
Yeah, I mentioned that in my review too.

glensc replied:
this might help:

```sh
rc=0
command1 || rc=1
command2 || rc=2
exit $rc
```

jrfnl commented:
@glensc @roelofr Thank you both for the reviews. You are right. (it was late when I was creating these fixes). I'll add another commit in a moment to sort this out.

In the mean time, I have already added two other extra commits with other fixes:
1. To allow for the test run to work on PHP 5.3 in anticipation of PR #51 being merged. See https://github.com/php-parallel-lint/PHP-Parallel-Lint/pull/51#issuecomment-808889253
2. I noticed that the Phar was quite large for this relatively small project. This was due to the Phar including all the `dev` dependencies, which shouldn't actually be included.
    I've fixed that now as well, resulting in the size of the generated Phar going down from 1.5 MB to 72 KB and still working as expected. 😁

jrfnl commented:
### Cs2pr for linting the Parallel Lint code (integration tests)

I've done a number of test runs with cs2pr for the integration tests and have in the end decided to remove the commit which added it.

Reason being: if there is a linting error in the Parallel Lint application, the checkstyle output will never get generated correctly anyway, so the cs2pr reporting would not work, no matter what.

Part of the reason for this is the current setup with all multiple classes in each file, causing parse errors to affect the application as a whole, even when the code containing the parse error would not be run.

It might be worth it to revisit this after PR #45 has been finished and merged, but for now, adding cs2pr for the linting of the code in this application itself would never work.


> And due to | filtering, the second command won't even trigger the error code to shell.
> note, cs2pr itself may trigger error if it finds error in input, haven't checked that.

@glensc It does, so that is not an issue.


### Test execution order

I also had another critical look at the execution order of the test commands based on the feedback.

> My first suggestion does not work, but setting the first test as `continue-on-error` does yield the requested results.

I agree with @roelofr's suggestion here, though have ended up taking it one step further.

Please see the commit description of commit https://github.com/php-parallel-lint/PHP-Parallel-Lint/pull/55/commits/194a080c8b794a5730611a321afac4b2eb884332 for full details.

I've also updated the PR description to match with the changes now made.

glensc commented:
https://github.com/php-parallel-lint/PHP-Parallel-Lint/pull/54 was merged, rebase time

glensc commented:
> As the system `php.ini` was not being used, no extensions were loaded, which was causing the errors.
> 
> This is a change that was introduced in Nette Tester 2.0.0. As of that version, you need to always tell the Nette testing framework explicitly which `php.ini` file to use, `-C` tells it to use the system-wide `php.ini`, with `-c` you can specify a path to a `php.ini` file and by default **_no `php.ini` is used_**.
> (_honestly, why did anyone ever think making that the default was a good idea ?!!?_)

I can speak from my personal experience, I've used `php -n` in my Linux Distribution, which has a highly modularized PHP, *every* extension being a shared module.

so, if you run your tests, you want to "remove" unwanted extensions, this is easy by skip loading `php.ini` after all, so `.ini` fragments with `extension=spl.so`, etc are not loaded, and you are left with minimal PHP modules. and if your tests run in a way that PHP interpreter is started for each test, this makes up a significant time-saver.

This approach did not work well, as it also disabled error reporting and other useful defaults.

And I think it's not relevant anymore, as you can fire up the container with PHP configured exactly for you. I.e the PHP environment is cattle, no longer a pet.

Search "Pets vs Cattle DevOps Concept" to read more.

This lead to for fixes for such environments:
- https://github.com/JakubOnderka/PHP-Parallel-Lint/pull/80
- https://github.com/phingofficial/phing/commit/8cfdd3ae293cb14c0e00816d6047d717153f6c9e (the ticket reference is trac, not github)

also, you can probably find up the nettle tester 2.0 commit and/or pull request/issue, which may be describes the reasoning behind, instead of speculating :)

jrfnl commented:
@glensc Oh, I totally agree that it is a good idea to have _the option_ to run without loading the `php.ini` file, it's just very contra-intuitive for that to be the default.

I did look at the commits when I was working on fixing it, but don't remember seeing the reasoning for it. Then again, I didn't search too hard as, more than anything, I was just glad to have found the cause of the issue & the solution to getting the tests running again.

ktomk commented:
>  it's just very contra-intuitive for that to be the default.

at the first time, it bites. but if you reduce the runtime environment it normally takes away side-effect problems en masse so after the first "hoppla the gap" it is then merely a straight forward ride and things then break in smaller increments, easy to handle in the gallop.

for cli utilities, another one is to strip down the environment and then see what happens. `env` is a good utility to do that. yes, it breaks things ;). but isn't testing all about breaking things?

containers? they need to be build first so should only be a second thought. system is already there, stripping the runtime is pretty viable and straight forward IMHO.
